Mathlib/Data/Nat/Choose/Sum.lean

```diff
@@ -128,6 +128,24 @@ theorem sum_Icc_choose (n k : ℕ) : ∑ m ∈ Icc k n, m.choose k = (n + 1).cho
       show Ico k (n + 1) = Icc k n by rfl, ih, choose_succ_succ' (n + 1)]
   · rw [choose_eq_zero_of_lt (by omega), Icc_eq_empty_of_lt h, sum_empty]
 
+/-- Summing `(i+k).choose k` for `i ∈ [0, n]` gives `(n + (k + 1)).choose (k + 1)`.
+
+Combinatorial interpretation: `(i+k).choose k` is the number of decompositions of `[0, i)` in
+`k + 1` (possibly empty) intervals (this follows from a stars and bars description). In particular,
+`(n + (k + 1)).choose (k + 1)` corresponds to decomposing `[0, n)` into `k + 2` intervals. By
+putting away the last interval (of some length `n-i`), we have to decompose the remaining interval
+`[0, i)` into `k+1` intervals, hence the sum.
+-/
+lemma sum_range_add_choose (k n : ℕ) :
+    ∑ i ∈ Finset.range (n + 1), (i + k).choose k = (n + (k + 1)).choose (k + 1) := by
+  induction n with
+  | zero => simp
+  | succ n ih =>
+      have : n + 1 + (k + 1) = (n + (k + 1)) + 1 := by omega
+      rw [this, choose_succ_succ', sum_range_succ, ← ih, add_comm]
+      congr 2
+      omega
```

Comment:
```suggestion
/-- **Zhu Shijie's identity** aka hockey-stick identity. Version with `range` instead of `Icc`. -/
lemma sum_range_add_choose (n k : ℕ) :
    ∑ i ∈ range (n + 1), (i + k).choose k = (n + k + 1).choose (k + 1) := by
  rw [← sum_Icc_choose (n + k) k, range_eq_Ico]
  convert (sum_map _ (addRightEmbedding k) (·.choose k)).symm using 2
  rw [map_add_right_Ico, zero_add, add_right_comm]; rfl
```

Reply:
In fact I've opened #15935 to add this lemma properly.

Reply:
Thanks for your shorter proof! However, you shouldn't have removed the combinatorial interpretation, because I'm pretty sure people looking for this lemma will grep for "stars and bars", so I'm adding it back.